pkg/network/bridge/plugin.go

```diff
@@ -0,0 +1,218 @@
+			s.Send(&pluginapi.ListAndWatchResponse{Devices: nbdp.DevicePlugin.Devs})
+		}
+	}
+}
+
+type Attachment struct {
+	DeviceID      string
+	ContainerPath string
+}
+
+func (nbdp *NetworkBridgeDevicePlugin) Allocate(ctx context.Context, r *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	var response pluginapi.AllocateResponse
+
+	for _, nic := range r.DevicesIDs {
+		dev := new(pluginapi.DeviceSpec)
+		attachmentPath := getAttachmentPath(nbdp.bridge, nic)
+		dev.HostPath = fakeDevicePath
+		dev.ContainerPath = attachmentPath
+		dev.Permissions = "r"
+		response.Devices = append(response.Devices, dev)
+		nbdp.attachmentCh <- &Attachment{
+			nic,
+			attachmentPath,
+		}
+	}
+
+	return &response, nil
+}
+
+func getAttachmentPath(bridge string, nic string) string {
+	return fmt.Sprintf("/tmp/device-plugin-network-bridge/%s/%s", bridge, nic)
+}
+
+func (nbdp *NetworkBridgeDevicePlugin) attachBridges() {
+	glog.V(3).Info("Running attachBridges procedure")
+
+	pendingAttachments := list.New()
+
+	cli, err := dockerutils.NewClient()
+	if err != nil {
+		glog.V(3).Info("Failed to connect to Docker")
+		panic(err)
+	}
+
+	for {
+		select {
+		case attachment := <-nbdp.attachmentCh:
+			glog.V(3).Infof("Received a new attachment request: %s", attachment)
+			pendingAttachments.PushBack(attachment)
+		default:
+			time.Sleep(time.Second)
+		}
+
+		for a := pendingAttachments.Front(); a != nil; a = a.Next() {
+			attachment := a.Value.(*Attachment)
+			glog.V(3).Infof("Handling pending attachment request for: %s", attachment.DeviceID)
+			containerID, err := cli.GetContainerIDByMountedDevice(attachment.ContainerPath)
+			if err != nil {
+				glog.V(3).Info("Container was not found")
+				continue
+			}
+
+			containerPid, err := cli.GetPidByContainerID(containerID)
+			if err != nil {
+				glog.V(3).Info("Failed to obtain container's pid")
+				continue
+			}
+
+			err = attachPodToBridge(nbdp.bridge, containerPid)
+			if err == nil {
+				glog.V(3).Info("Successfully attached pod to a bridge")
+			} else {
+				glog.V(3).Infof("Pod attachment failed with: %s", err)
+			}
+			pendingAttachments.Remove(a)
+		}
+	}
+}
+
+func attachPodToBridge(bridgeName string, containerPid int) error {
+	linkName := randIfaceName()
+	peerName := randIfaceName()
+
+	bridge, err := netlink.LinkByName(bridgeName)
+	if err != nil {
+		return err
+	}
+
+	link := &netlink.Veth{LinkAttrs: netlink.LinkAttrs{Name: linkName, MasterIndex: bridge.Attrs().Index}, PeerName: peerName}
+
+	err = netlink.LinkAdd(link)
+	if err != nil {
+		return err
+	}
+
+	peer, err := netlink.LinkByName(peerName)
+	if err != nil {
+		netlink.LinkDel(link)
+		return err
+	}
+
+	err = netlink.LinkSetNsPid(peer, containerPid)
+	if err != nil {
+		netlink.LinkDel(link)
+		return err
+	}
+
+	return nil
+}
+
+const (
+	ifaceNameLen    = 15
+	ifaceNamePrefix = "nic_"
+	letterBytes     = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+)
+
+func randIfaceName() string {
```

Comment:
Nit: randInterfaceName (no reason to shorten Interface to Iface).

Reply:
Done